### PR TITLE
Add timeout parameter to `DockerOperator`

### LINKED ISSUE #22115: add timeout to DockerOperator
### Body

`APIClient` has [timeout](https://github.com/docker/docker-py/blob/b27faa62e792c141a5d20c4acdd240fdac7d282f/docker/api/client.py#L84) param which sets default timeout for API calls. The package [default](https://github.com/docker/docker-py/blob/7779b84e87bea3bac3a32b3ec1511bc1bfaa44f1/docker/constants.py#L6) is 60 seconds which may not be enough for some process.

The needed fix is to allow to set `timeout` to the APIClient constructor in DockerHook and DockerOperator:

for example in:
https://github.com/apache/airflow/blob/05f3a309668288e03988fc4774f9c801974b63d0/airflow/providers/docker/hooks/docker.py#L89

Note that in some cases DockerOperator uses APIClient directly for example in:
https://github.com/apache/airflow/blob/05f3a309668288e03988fc4774f9c801974b63d0/airflow/providers/docker/operators/docker.py#L390

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by Bowrna**
Included the timeout support in APIClient of Docker in DockerOperator and Hook under Docker provider. 
closes: #22115
related:  #22115

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by Bowrna** [comment]
While I have added timeout as specified in APIClient, I am not sure about adding test cases in the current test cases.

If anybody could suggest changes that I have to make in the test, it would be useful. 

https://github.com/apache/airflow/blob/51d61df5a656101046a7825be53ac61ac4f2b047/tests/providers/docker/hooks/test_docker.py

File: airflow/providers/docker/hooks/docker.py
Comment (eladkal) [inline airflow/providers/docker/hooks/docker.py]: can/should we import the default of the package rather than hard code value from our side?

https://github.com/docker/docker-py/blob/a48a5a9647761406d66e8271f19fab7fa0c5f582/docker/constants.py#L6

Comment (Bowrna) [reply airflow/providers/docker/hooks/docker.py]: @eladkal yes I could import the default from the package, that way it's easy to maintain. One other option I thought is to keep it None by default and not pass it at all if it's None. So that it will pick the default one when we don't pass to APIClient

Comment (Bowrna) [reply airflow/providers/docker/hooks/docker.py]: @eladkal which one do you think will work best? Keeping None as default or passing the default of package by importing?

Comment (eladkal) [reply airflow/providers/docker/hooks/docker.py]: If setting to None falls to the default within the package then I guess it's prefered.

Comment (Bowrna) [reply airflow/providers/docker/hooks/docker.py]: Ok @eladkal then i will verify it and make relevant changes

Comment (Bowrna) [reply airflow/providers/docker/hooks/docker.py]: @eladkal I have added default within package by importing it as default param.

**Review by pingzh — COMMENTED** [review]
nice 👍

**Review by eladkal — APPROVED** [review]
LGTM
will merge when CI is green

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Some static checks/tests fail :).

**Comment by Bowrna** [comment]
> Some static checks/tests fail :).

Fixing it @potiuk

**Comment by Bowrna** [comment]
Static checks are fixed. While Postgres10 and Sqlite tests still fail. I am not sure how to fix those.

**Comment by eladkal** [comment]
It fails because there are assertion errors:
https://github.com/apache/airflow/runs/5710010118?check_suite_focus=true#step:9:7584

Something with how we manage the default value isn't quite right (I think)

**Comment by potiuk** [comment]
No. Default is fine. The tests have to be fixed because they don't expec the timeout parameter to be passed.

**Comment by potiuk** [comment]
That's the case were tests **should** be changed reflect the new reality after the change :)

**Comment by Bowrna** [comment]
I see, let me fix the tests @potiuk @eladkal

File: tests/providers/docker/operators/test_docker.py
Comment (eladkal) [inline tests/providers/docker/operators/test_docker.py]:
```suggestion
from docker.constants import DEFAULT_TIMEOUT_SECONDS
from docker.errors import APIError
```

to fix static checks

**Comment by Bowrna** [comment]
@potiuk @eladkal  I am trying to test this feature by creating a dag with DockerOperator and timeout

```
from datetime import datetime, timedelta

from airflow import DAG
from airflow.operators.bash import BashOperator
from airflow.providers.docker.operators.docker import DockerOperator

dag = DAG(
    'docker_sample',
    default_args={'retries': 1},
    schedule_interval=timedelta(minutes=10),
    start_date=datetime(2021, 1, 1),
    catchup=False,
)

t1 = BashOperator(task_id='print_date', bash_command='date', dag=dag)

t2 = BashOperator(task_id='sleep', bash_command='sleep 5', retries=3, dag=dag)

t3 = DockerOperator(
    docker_url='unix:///var/run/docker.sock',  # Set your docker URL
    command='/bin/sleep 30',
    image='centos:latest',
    network_mode='bridge',
    task_id='docker_op_tester',
    dag=dag,
    timeout=10,
)

t4 = BashOperator(task_id='print_hello', bash_command='echo "hello world!!!"', dag=dag)


t1 >> t2
t1 >> t3
t3 >> t4
```
The above task runs successfully. I am not sure if the task t3 is supposed to run successfully because I have set the command to execute for 30 seconds but the timeout is 10 seconds. Do I have to expect it to throw an error? 
Any review about this would be helpful for me to learn more about DockerOperator

**Comment by potiuk** [comment]
I think `timeout` is not what you think it is. It's not well described in the docs (might be a good idea to add docstrings) :). - is a timeout for setting up the connection with Docker via HTTP.  You should take a look how it is passed to Hook and ClientAPI https://docker-py.readthedocs.io/en/1.1.0/api/ .

What you really wanted there is "execution_timeout" from the BaseOperator: https://airflow.apache.org/docs/apache-airflow/1.10.6/_api/airflow/models/index.html#airflow.models.BaseOperator

**Pull request merged**
